src/infra/core/src/dependency_graph_service_inmem.rs

```diff
@@ -193,6 +194,100 @@ impl DependencyGraphService for DependencyGraphServiceInMemory {
             .await
     }
 
+    async fn get_all_upstream_dependencies(
+        &self,
+        dataset_ids: Vec<DatasetID>,
+    ) -> Result<DatasetIDStream, GetUpstreamDependenciesError> {
+        self.ensure_datasets_initially_scanned()
+            .await
+            .int_err()
+            .map_err(GetUpstreamDependenciesError::Internal)
+            .unwrap();
+
+        let state = self.state.lock().await;
+
+        let nodes_to_search = dataset_ids
+            .iter()
+            .map(|dataset_id| {
+                state
+                    .get_dataset_node(dataset_id)
+                    .map_err(GetUpstreamDependenciesError::DatasetNotFound)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let mut result = Vec::new();
+        let mut visited_indexes = HashSet::new();
+
+        let mut reversed_graph = state.datasets_graph.clone();
+
+        reversed_graph.reverse();
+
+        for node_index in &nodes_to_search {
+            let mut bfs = Bfs::new(&reversed_graph, *node_index);
+            // Push passed node to result
+            result.push(reversed_graph.node_weight(*node_index).unwrap().clone());
+            visited_indexes.insert(*node_index);
+
+            while let Some(current_node_index) = bfs.next(&reversed_graph) {
+                if &current_node_index != node_index
+                    && !visited_indexes.contains(&current_node_index)
+                {
+                    visited_indexes.insert(*node_index);
+                    result.push(reversed_graph.node_weight(*node_index).unwrap().clone());
+                }
+            }
+        }
+
+        Ok(Box::pin(tokio_stream::iter(result)))
+    }
+
+    async fn get_all_downstream_dependencies(
+        &self,
+        dataset_ids: Vec<DatasetID>,
+    ) -> Result<DatasetIDStream, GetDownstreamDependenciesError> {
+        self.ensure_datasets_initially_scanned()
+            .await
+            .int_err()
+            .map_err(GetDownstreamDependenciesError::Internal)
+            .unwrap();
+
+        let state = self.state.lock().await;
+
+        let nodes_to_search = dataset_ids
+            .iter()
+            .map(|dataset_id| {
+                state
+                    .get_dataset_node(dataset_id)
+                    .map_err(GetDownstreamDependenciesError::DatasetNotFound)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        let mut node_indexes_result = HashSet::new();
+        let mut result = vec![];
+
+        for node_index in &nodes_to_search {
+            let mut current_node_result = vec![];
+            depth_first_search(&state.datasets_graph, Some(*node_index), |event| {
```

Comment:
DFS implementation seems to accept multiple start nodes, did you try this option?
It is also has a built-in functionality to ignore duplicate visits.

Reply:
Yep. I've tried it before but faced issue, that node we are searching for always at the begging, found good solution for that and changed it!